File: pset6/legibilidade/readability.py
from cs50 import get_string


def count_letters_words_sentences(text):
    letters = 0
    words = 1
    sentences = 0

    for char in text:
        if char.isalpha():
            letters += 1
        elif char.isspace():
            words += 1
        elif char in ['.', '!', '?']:
            sentences += 1

    return letters, words, sentences


def calculate_index(letters, words, sentences):
    L = (letters / words) * 100
    S = (sentences / words) * 100
    return 0.0588 * L - 0.296 * S - 15.8


def main():
    text = get_string("Text: ")
    letters, words, sentences = count_letters_words_sentences(text)
    index = calculate_index(letters, words, sentences)

    print(f"Index: {index:.2f}")

    if index < 1:
        print("Less than Grade 1")
    elif index > 16:
        print("Grade 16+")
    else:
        print(f"Grade {round(index)}")


if __name__ == "__main__":
    main()

File: C/readability.c
#include <cs50.h>
#include <ctype.h>
#include <math.h>
#include <stdio.h>
#include <string.h>

void countLettersWordsSentences(string text, int *letters, int *words, int *sentences);
float calculateIndex(int letters, int words, int sentences);

int main(void)
{
    string text = get_string("Text: ");
    int letters = 0, words = 1, sentences = 0;

    countLettersWordsSentences(text, &letters, &words, &sentences);

    float index = calculateIndex(letters, words, sentences);

    printf("Index: %.2f\n", index);

    if (index < 1)
    {
        printf("Less than Grade 1\n");
    }
    else if (index > 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %i\n", (int) round(index));
    }
}

void countLettersWordsSentences(string text, int *letters, int *words, int *sentences)
{
    for (int i = 0, n = strlen(text); i < n; i++)
    {
        if (isalpha(text[i]))
        {
            (*letters)++;
        }
        else if (isspace(text[i]))
        {
            (*words)++;
        }
        else if (text[i] == '.' || text[i] == '!' || text[i] == '?')
        {
            (*sentences)++;
        }
    }
}

float calculateIndex(int letters, int words, int sentences)
{
    float L = (float) letters / words * 100;
    float S = (float) sentences / words * 100;
    return 0.0588 * L - 0.296 * S - 15.8;
}
